File: ghrah-observer-webui/packages/protocol/src/payloads/room.ts
import { z } from "zod";

// ─── Room Payloads ───

export const RoomSubjectTypeSchema = z.enum(["agent", "human"]);

export const RoomStatusSchema = z.enum(["active", "archived"]);

export const RoomMemberSchema = z.object({
  subject: z.string(),
  subject_type: RoomSubjectTypeSchema,
  /** 可读名称；subject 对 agent 成员保存 project-scoped 稳定 ID。 */
  subject_name: z.string().optional(),
  joined_at: z.string().optional().default(""),
});

export const RoomInfoPayloadSchema = z.object({
  room_id: z.string(),
  project_id: z.string(),
  name: z.string(),
  status: RoomStatusSchema.optional().default("active"),
  members: z.array(RoomMemberSchema).optional().default([]),
  seq_watermark: z.number().int().optional().default(0),
  version: z.number().int().optional().default(1),
  created_at: z.string().optional().default(""),
  updated_at: z.string().optional().default(""),
  archived_at: z.string().nullable().optional().default(null),
});

export const RoomLogEntryPayloadSchema = z.object({
  id: z.string(),
  room_id: z.string(),
  seq: z.number().int(),
  author: z.string(),
  author_type: RoomSubjectTypeSchema,
  timestamp: z.number(),
  data: z.record(z.unknown()).optional().default({}),
});

export const RoomCreatePayloadSchema = z.object({
  project_id: z.string(),
  name: z.string(),
});

export const RoomListPayloadSchema = z.object({
  project_id: z.string().nullable().optional(),
  status: RoomStatusSchema.nullable().optional(),
});

export const RoomIdPayloadSchema = z.object({
  room_id: z.string(),
});

export const RoomUpdatePayloadSchema = z.object({
  room_id: z.string(),
  name: z.string().nullable().optional(),
  expected_version: z.number().int().nullable().optional(),
});

export const RoomLifecyclePayloadSchema = z.object({
  room_id: z.string(),
  expected_version: z.number().int(),
});

export const RoomDeletePayloadSchema = RoomLifecyclePayloadSchema;

export const RoomJoinPayloadSchema = z.object({
  room_id: z.string(),
  subject: z.string(),
  subject_type: RoomSubjectTypeSchema,
  subject_name: z.string().optional(),
});

export const RoomLeavePayloadSchema = z.object({
  room_id: z.string(),
  subject: z.string(),
});

export const RoomGetLogPayloadSchema = z.object({
  room_id: z.string(),
  since_seq: z.number().int().nullable().optional(),
  limit: z.number().int().optional().default(100),
});

export const RoomSendPayloadSchema = z.object({
  room_id: z.string(),
  author: z.string(),
  author_type: RoomSubjectTypeSchema,
  data: z.record(z.unknown()).optional().default({}),
});

export const RoomListResultPayloadSchema = z.object({
  rooms: z.array(RoomInfoPayloadSchema).optional().default([]),
  count: z.number().int().optional().default(0),
});

export const RoomLogResultPayloadSchema = z.object({
  entries: z.array(RoomLogEntryPayloadSchema).optional().default([]),
  count: z.number().int().optional().default(0),
});

// ─── Room Event Payloads ───

export const RoomEventPayloadSchema = z.object({
  room: RoomInfoPayloadSchema,
});

export const RoomDeletedEventPayloadSchema = z.object({
  room_id: z.string(),
  project_id: z.string(),
});

export const RoomMemberEventPayloadSchema = z.object({
  room: RoomInfoPayloadSchema,
  member: RoomMemberSchema.nullable().optional(),
  subject: z.string().nullable().optional(),
});

export const RoomLogEventPayloadSchema = z.object({
  entry: RoomLogEntryPayloadSchema,
});

export type RoomSubjectType = z.infer<typeof RoomSubjectTypeSchema>;
export type RoomStatus = z.infer<typeof RoomStatusSchema>;
export type RoomMember = z.infer<typeof RoomMemberSchema>;
export type RoomInfoPayload = z.infer<typeof RoomInfoPayloadSchema>;
export type RoomLogEntryPayload = z.infer<typeof RoomLogEntryPayloadSchema>;
export type RoomCreatePayload = z.infer<typeof RoomCreatePayloadSchema>;
export type RoomListPayload = z.infer<typeof RoomListPayloadSchema>;
export type RoomIdPayload = z.infer<typeof RoomIdPayloadSchema>;
export type RoomUpdatePayload = z.infer<typeof RoomUpdatePayloadSchema>;
export type RoomLifecyclePayload = z.infer<typeof RoomLifecyclePayloadSchema>;
export type RoomDeletePayload = z.infer<typeof RoomDeletePayloadSchema>;
export type RoomJoinPayload = z.infer<typeof RoomJoinPayloadSchema>;
export type RoomLeavePayload = z.infer<typeof RoomLeavePayloadSchema>;
export type RoomGetLogPayload = z.infer<typeof RoomGetLogPayloadSchema>;
export type RoomSendPayload = z.infer<typeof RoomSendPayloadSchema>;
export type RoomListResultPayload = z.infer<typeof RoomListResultPayloadSchema>;
export type RoomLogResultPayload = z.infer<typeof RoomLogResultPayloadSchema>;
export type RoomEventPayload = z.infer<typeof RoomEventPayloadSchema>;
export type RoomDeletedEventPayload = z.infer<typeof RoomDeletedEventPayloadSchema>;
export type RoomMemberEventPayload = z.infer<typeof RoomMemberEventPayloadSchema>;
export type RoomLogEventPayload = z.infer<typeof RoomLogEventPayloadSchema>;

File: ghrah-protocol/src/ghrah/protocol/payloads/room.py
# SPDX-FileCopyrightText: 2026 chenxya <chenxya@ghrah.org>
#
# SPDX-License-Identifier: Apache-2.0
"""Room 域载荷模型（成员、命令、结果与事件）。"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from ghrah.protocol.enums import RoomStatus, RoomSubjectType

# ─── Room 载荷模型 ───


class RoomMember(BaseModel):
    """Room 成员（多对多：agent / human 均可为成员）。"""

    subject: str
    subject_type: RoomSubjectType
    subject_name: str = ""
    joined_at: str = ""


class RoomInfoPayload(BaseModel):
    """Room 信息载荷（room_get/room_list 结果项及 ROOM_* 事件回执）。

    对齐 ghrah-subject room/models.RoomRecord；seq_watermark 为该 room
    已分配的最大 RoomLog seq（Subject 单点分配）。
    """

    room_id: str
    project_id: str
    name: str
    status: RoomStatus = RoomStatus.ACTIVE
    members: list[RoomMember] = Field(default_factory=list)
    seq_watermark: int = 0
    version: int = 1
    created_at: str = ""
    updated_at: str = ""
    archived_at: str | None = None


class RoomLogEntryPayload(BaseModel):
    """RoomLog 节点载荷 = 最小信封 + 自由 map（不预置业务枚举 schema）。"""

    id: str
    room_id: str
    seq: int
    author: str
    author_type: RoomSubjectType
    timestamp: float
    data: dict[str, Any] = Field(default_factory=dict)


class RoomCreatePayload(BaseModel):
    """room_create 命令载荷。"""

    project_id: str
    name: str


class RoomListPayload(BaseModel):
    """room_list 命令载荷（可选 project/status 过滤）。"""

    project_id: str | None = None
    status: RoomStatus | None = None


class RoomIdPayload(BaseModel):
    """room_get / room_update / room_delete / room_get_members /
    room_get_log / room_send 等命令的 room_id 键控基类。"""

    room_id: str


class RoomUpdatePayload(BaseModel):
    """room_update 命令载荷（乐观锁）。"""

    room_id: str
    name: str | None = None
    expected_version: int | None = None


class RoomLifecyclePayload(RoomIdPayload):
    """room_archive / room_restore 命令载荷（乐观锁）。"""

    expected_version: int


class RoomDeletePayload(RoomLifecyclePayload):
    """room_delete 命令载荷。"""


class RoomJoinPayload(BaseModel):
    """room_join 命令载荷。"""

    room_id: str
    subject: str
    subject_type: RoomSubjectType
    subject_name: str = ""


class RoomLeavePayload(BaseModel):
    """room_leave 命令载荷。"""

    room_id: str
    subject: str


class RoomGetLogPayload(BaseModel):
    """room_get_log 命令载荷（since_seq 增量 / limit 截尾）。"""

    room_id: str
    since_seq: int | None = None
    limit: int = 100


class RoomSendPayload(BaseModel):
    """room_send 命令载荷（人类/Mock 公开路径；agent 路径由 Core send
    ability 解析后经同一命令面收敛，双调用方分流）。"""

    room_id: str
    author: str
    author_type: RoomSubjectType
    data: dict[str, Any] = Field(default_factory=dict)


class RoomListResultPayload(BaseModel):
    """room_list 命令结果载荷。"""

    rooms: list[RoomInfoPayload] = Field(default_factory=list)
    count: int = 0


class RoomLogResultPayload(BaseModel):
    """room_get_log 命令结果载荷。"""

    entries: list[RoomLogEntryPayload] = Field(default_factory=list)
    count: int = 0  # ─── Room 事件载荷模型 ───


class RoomEventPayload(BaseModel):
    """room_created / room_updated 事件载荷（全量快照）。"""

    room: RoomInfoPayload


class RoomDeletedEventPayload(BaseModel):
    """room_deleted 事件载荷。"""

    room_id: str
    project_id: str


class RoomMemberEventPayload(BaseModel):
    """room_member_joined / room_member_left 事件载荷。

    joined 携带 member；left 仅 subject（成员已移出全量快照）。
    """

    room: RoomInfoPayload
    member: RoomMember | None = None
    subject: str | None = None


class RoomLogEventPayload(BaseModel):
    """room_log_appended 事件载荷。"""

    entry: RoomLogEntryPayload
